File: Sync.js
const uuid = require('uuid');

// Define constants
const DATA_KEYS = ["a", "b", "c"];

// Device class
class Device {
    constructor(id) {
        this._id = id;
        this.records = [];
        this.sent = [];
    }

    obtainData() {
        if (Math.random() < 0.4) {
            return {};
        }

        const record = {
            type: 'record',
            timestamp: new Date().toISOString(),
            dev_id: this._id,
            data: DATA_KEYS.reduce((acc, key) => {
                acc[key] = uuid.v4();
                return acc;
            }, {})
        };
        this.sent.push(record);
        return record;
    }

    probe() {
        if (Math.random() < 0.5) {
            return {};
        }

        return {
            type: 'probe',
            dev_id: this._id,
            from: this.records.length
        };
    }

    onMessage(data) {
        console.log(`Device ID: ${this._id}, Message Received:`, data);
        if (!data || Math.random() < 0.6) {
            // Check if data is undefined or randomly skip processing
            return;
        }
    
        if (data.type === 'update') {
            console.log('Handling update message:', data);
            const from = data.from;
            if (from > this.records.length) {
                return;
            }
            this.records = this.records.slice(0, from).concat(data.data);
        }
    }
    
}

// SyncService class
class SyncService {
    constructor() {
        this.serverRecords = [];
    }

    onMessage(data) {
        if (data.type === 'probe') {
            const from = data.from;
            const updateData = {
                type: 'update',
                from: from,
                data: this.serverRecords.slice(from)
            };
            return updateData;
        }
        // No return value required for handling 'record' type
    }
}

// Test function
function testSyncing() {
    console.log("Starting testSyncing");
    const devices = Array.from({ length: 10 }, (_, i) => new Device(`dev_${i}`));
    const syn = new SyncService();

    const N = 1000;
    for (let i = 0; i < N; i++) {
        console.log(`Iteration: ${i + 1}`);
        for (const dev of devices) {
            console.log(`Device ID: ${dev._id}`);
            syn.onMessage(dev.obtainData());
            dev.onMessage(syn.onMessage(dev.probe()));
        }
    }

    let done = false;
    while (!done) {
        for (const dev of devices) dev.onMessage(syn.onMessage(dev.probe()));
        const num_recs = devices[0].records.length;
        done = devices.every((_dev) => _dev.records.length === num_recs);
    }

    const ver_start = Array(devices.length).fill(0);
    for (let i = 0; i < devices[0].records.length; i++) {
        const rec = devices[0].records[i];
        const devIdx = parseInt(rec.dev_id.split("_").pop());
        assertEquivalent(rec, devices[devIdx].sent[ver_start[devIdx]]);
        for (const _dev of devices.slice(1)) {
            assertEquivalent(rec, _dev.records[i]);
        }
        ver_start[devIdx] += 1;
    }
}

function assertEquivalent(d1, d2) {
    if (d1.dev_id !== d2.dev_id || d1.timestamp !== d2.timestamp) {
        throw new Error("Assertion failed: Records are not equivalent");
    }

    for (const key of DATA_KEYS) {
        if (d1.data[key] !== d2.data[key]) {
            throw new Error("Assertion failed: Records are not equivalent");
        }
    }
}

// Execute the test
testSyncing();

File: Python/Sync.py
import uuid
import random
from datetime import datetime

# Define constants
DATA_KEYS = ["a", "b", "c"]

# Device class
class Device:
    def __init__(self, id):
        self._id = id
        self.records = []
        self.sent = []

    def obtainData(self):
        if random.random() < 0.4:
            return {}
        record = {
            "type": "record",
            "timestamp": datetime.now().isoformat(),
            "dev_id": self._id,
            "data": {key: str(uuid.uuid4()) for key in DATA_KEYS}
        }
        self.sent.append(record)
        return record

    def probe(self):
        if random.random() < 0.5:
            return {}
        return {
            "type": "probe",
            "dev_id": self._id,
            "from": len(self.records)
        }

    def onMessage(self, data):
        print(f"Device ID: {self._id}, Message Received:", data)
        if not data or random.random() < 0.6:
            # Check if data is empty or randomly skip processing
            return
        if data["type"] == "update":
            print("Handling update message:", data)
            from_index = data["from"]
            if from_index > len(self.records):
                return
            self.records = self.records[:from_index] + data["data"]

# SyncService class
class SyncService:
    def __init__(self):
        self.serverRecords = []

    def onMessage(self, data):
        if data["type"] == "probe":
            from_index = data["from"]
            updateData = {
                "type": "update",
                "from": from_index,
                "data": self.serverRecords[from_index:]
            }
            return updateData
        # No return value required for handling 'record' type

# Test function
def testSyncing():
    print("Starting testSyncing")
    devices = [Device(f"dev_{i}") for i in range(10)]
    syn = SyncService()
    N = 1000
    for i in range(N):
        print(f"Iteration: {i + 1}")
        for dev in devices:
            print(f"Device ID: {dev._id}")
            syn.onMessage(dev.obtainData())
            dev.onMessage(syn.onMessage(dev.probe()))
    done = False
    while not done:
        for dev in devices:
            dev.onMessage(syn.onMessage(dev.probe()))
        num_recs = len(devices[0].records)
        done = all(dev.records.length == num_recs for dev in devices)
    ver_start = [0] * len(devices)
    for i in range(len(devices[0].records)):
        rec = devices[0].records[i]
        devIdx = int(rec["dev_id"].split("_").pop())
        assertEquivalent(rec, devices[devIdx].sent[ver_start[devIdx]])
        for _dev in devices[1:]:
            assertEquivalent(rec, _dev.records[i])
        ver_start[devIdx] += 1

def assertEquivalent(d1, d2):
    if d1["dev_id"] != d2["dev_id"] or d1["timestamp"] != d2["timestamp"]:
        raise Exception("Assertion failed: Records are not equivalent")
    for key in DATA_KEYS:
        if d1["data"][key] != d2["data"][key]:
            raise Exception("Assertion failed: Records are not equivalent")

# Execute the test
testSyncing()


